errors/middleware-upgrade-guide.md

```diff
@@ -0,0 +1,290 @@
+
+  return response
+}
+```
+
+#### After
+
+```typescript
+// middleware.ts
+export function middleware() {
+  const response = new NextResponse()
+
+  // set a cookie
+  response.cookies.set('vercel', 'fast')
+
+  // set another cookie with options
+  response.cookies.set('nextjs', 'awesome', { path: '/test' })
+
+  // get all the details of a cookie
+  const [value, options] = response.cookies.getWithOptions('vercel')
+  console.log(value) // => 'fast'
+  console.log(options) // => { Path: '/test' }
+
+  // delete a cookie means mark it as expired
+  response.cookies.delete('vercel')
+
+  // clear all cookies means mark all of them as expired
+  response.cookies.clear()
+}
+```
+
+## No More Page Match Data
+
+### Summary of changes
+
+- Use [`URLPattern`](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) to check if a Middleware is being invoked for a certain page match
+
+### Explanation
+
+Currently, Middleware estimates whether you are serving an asset of a Page based on the Next.js routes manifest (internal configuration). This value is surfaced through `request.page`.
+
+To make this more accurate, we are recommending to use the web standard `URLPattern` API.
+
+### How to upgrade
+
+Use [`URLPattern`](https://developer.mozilla.org/en-US/docs/Web/API/URLPattern) to check if a Middleware is being invoked for a certain page match.
+
+#### Before
+
+```typescript
+// middleware.ts
+import { NextRequest, NextResponse } from 'next/server'
+
+export function middleware(request: NextRequest) {
+  const { params } = event.request.page
+  const { locale, slug } = params
+
+  if (locale && slug) {
+    const { search, protocol, host } = request.nextUrl
+    const url = new URL(`${protocol}//${locale}.${host}/${slug}${search}`)
+    return NextResponse.redirect(url)
+  }
+}
+```
+
+#### After
+
+```typescript
+// middleware.ts
+import { NextRequest, NextResponse } from 'next/server'
+
+const PATTERNS = [
+  [
+    new URLPattern({ pathname: '/:locale/:slug' }),
+    ({ pathname }) => pathname.groups,
+  ],
+]
+
+const params = (url) => {
+  const input = url.split('?')[0]
+  let result = {}
+
+  for (const [pattern, handler] of PATTERNS) {
+    const patternResult = pattern.exec(input)
+    if (patternResult !== null && 'pathname' in patternResult) {
+      result = handler(patternResult)
+      break
+    }
+  }
+  return result
+}
+
+export function middleware(request: NextRequest) {
+  const { locale, slug } = params(request.url)
+
+  if (locale && slug) {
+    const { search, protocol, host } = request.nextUrl
+    const url = new URL(`${protocol}//${locale}.${host}/${slug}${search}`)
+    return NextResponse.redirect(url)
+  }
+}
+```
+
+## Executing Middleware on Internal Next.js Requests
+
+### Summary of changes
+
+- Middleware will be executed for _all_ requests, including `_next`
+
+### Explanation
+
+Currently, we do not execute Middleware for `_next` requests, due the authorization use cases.
```

Comment:
I had to re-read that section twice. I think changing this to "So far, we did not..." could make it clearer?